### ISSUE TITLE
400 less fps when drawing circles/arcs in 3.3

### ISSUE THREAD

**Issue by rune56** [godotengine/godot#47824]
**Godot version:**
Godot_v3.3-rc8_win64
i also tested 3.2.4 rc1 (both win32 and win64) and same issue
(no issues in 3.2.3, 3.2.2 or 3.2.1)

**OS/device including version:**
windows 10
gtx 1080
gles2 and gles3

**Issue description:**
drawing circles/arcs lowers fps a lot more than it does in 3.2.3
using the repro i get **104 fps in 3.3** and **519 in 3.2.3**

i tried drawing rects and lines and it worked fine not sure about anything else
also my projects that dont draw circles are about the same fps and in one case i even got a 1500 fps increase using 3.3

**Steps to reproduce:**
```
func _draw():
	for i in 200:
		draw_circle(Vector2(0, 0), 4.0, Color(1,0,0))
		draw_arc(Vector2(0,0), 4.0, 0, 7, 32, Color(0,0,0), 1.0, false)
```

**Minimal reproduction project:**
[repro.zip](https://github.com/godotengine/godot/files/6295553/repro.zip)

**a solution:**
out of curiosity i tried the (outdated) [circle arc methods from the docs](https://docs.godotengine.org/en/latest/tutorials/2d/custom_drawing_in_2d.html) which gives me **1140 fps in 3.3** and **186 in 3.2.3**
the only problem is that the arc doesnt have aa (although irrelevant to this issue)

heres that if you want to compare or something
```
func draw_circle_arc(center, radius, angle_from, angle_to, color):
	var nb_points = 8
	var points_arc = PoolVector2Array()

	for i in range(nb_points + 1):
		var angle_point = deg2rad(angle_from + i * (angle_to-angle_from) / nb_points - 90)
		points_arc.push_back(center + Vector2(cos(angle_point), sin(angle_point)) * radius)

	for index_point in range(nb_points):
		draw_line(points_arc[index_point], points_arc[index_point + 1], color)

func draw_circle_arc_poly(center, radius, angle_from, angle_to, color):
	var nb_points = 32
	var points_arc = PoolVector2Array()
	points_arc.push_back(center)
	var colors = PoolColorArray([color])
	
	for i in range(nb_points + 1):
		var angle_point = deg2rad(angle_from + i * (angle_to - angle_from) / nb_points - 90)
		points_arc.push_back(center + Vector2(cos(angle_point), sin(angle_point)) * radius)
	draw_polygon(points_arc, colors)

func _draw():
	for i in 200:
		draw_circle_arc_poly(Vector2(0,0), 4.0, 0, 360, Color(1,0,0))
		draw_circle_arc(Vector2(0,0), 4.0, 0, 360, Color(0,0,0))
```

**Comment by akien-mga** [comment]
This seems to be device/driver-specific as I can't reproduce the issue on Linux with Intel HD Graphics 630 and AMD Radeon RX Vega M. Details below.

The main change between 3.2.3 and 3.3 which might affect this is the extension of 2D batching to GLES3 (in 3.2.3 batching was only for GLES2).

Can you see if you get different results if you turn off batching in the project settings? (Rendering > Batching). There are also various options you can play with to see if they make a difference.

----

### Test results

Note that to avoid having a performance hit from running the editor itself, I'm running the project directly from the command line, not from the editor.

```
$ inxi -CSG
System:    Host: cauldron Kernel: 5.11.13-desktop-2.mga9 x86_64 bits: 64 Desktop: KDE Plasma 5.21.3 Distro: Mageia 9 mga9 
CPU:       Info: Quad Core model: Intel Core i7-8705G bits: 64 type: MT MCP cache: L2: 8 MiB 
           Speed: 3590 MHz min/max: 800/4100 MHz Core speeds (MHz): 1: 3590 2: 3863 3: 3775 4: 3840 5: 3837 6: 3894 7: 3880 
           8: 3860 
Graphics:  Device-1: Intel HD Graphics 630 driver: i915 v: kernel 
           Device-2: Advanced Micro Devices [AMD/ATI] Polaris 22 XL [Radeon RX Vega M GL] driver: amdgpu v: kernel 
           Device-3: Cheng Uei Precision Industry (Foxlink) HP Wide Vision FHD Camera type: USB driver: uvcvideo 
           Display: x11 server: Mageia X.org 1.20.10 driver: loaded: intel,v4l resolution: 1920x1080~60Hz 
           OpenGL: renderer: Mesa Intel HD Graphics 630 (KBL GT2) v: 4.6 Mesa 21.0.2
```

I seem to get more or less the same metrics with both GLES2 and GLES3 (maybe +20 FPS for AMD with GLES2 on both 3.2.3 and 3.3 RC 8 compared to GLES3).

| GPU       | 3.2.3 stable | 3.3 RC 8   |
|-----------|--------------|------------|
| Intel     | 210 FPS      | 380 FPS    |
| AMD       | 400 FPS      | 530 FPS    |

**Comment by akien-mga** [comment]
Testing on Windows 10 with my laptop with the AMD Radeon RX Vega 56, I seem to be able to reproduce a performance drop in 3.3 RC 8 compared to 3.2.3 stable, though both give me surprisingly bad performance metrics on this repo, compared to Linux drivers on the same hardware.

| GPU       | 3.2.3 stable | 3.3 RC 8   |
|-----------|--------------|------------|
| AMD       | 188 FPS      | 80 FPS    |

I'm not so familiar with debugging GPU drivers on Windows, but I can see in the Task Manager's Performance tab that the GPU is 100% busy on 3.2.3 stable, and only 66% busy on 3.3 RC 8. In both releases the CPU is only 5% busy.

Disabling batching doesn't seem to make a difference.

Here's some performance metrics from Radeon Software:

![image](https://user-images.githubusercontent.com/4701338/114392983-ca398180-9b99-11eb-98c1-9cf59a942c9e.png)

**Comment by pycbouh** [comment]
I can confirm similar results on another AMD GPU, 315 fps vs 170 fps on average. Interestingly enough, the number of draw calls is the same for both versions, and there is no output in the console from batching doing anything. So my assumption is just general degradation for some reason.

Frame times tell a different story. In 3.2.3 lows are lower, but highs tend to be higher. It gets from 2.9-3 ms to 7 ms on average and up to 10+ ms sometimes. In 3.3 the averages are more stable, it flactuates from 5.9 ms to 7+ ms, with no outliers in either direction.

![image](https://user-images.githubusercontent.com/11782833/114397117-01625f80-9ba7-11eb-8d93-7d96b51a0fec.png)
![image](https://user-images.githubusercontent.com/11782833/114397147-07f0d700-9ba7-11eb-9054-583c5eb4e8cd.png)

Radeon 5700 XT, driver version 21.3.1.

**Comment by lawnjelly** [comment]
Note that neither circles or arcs (polyline) are batched, so there shouldn't be a huge amount of differences codewise. It's possible that an extra state being set (possibly to fix a bug) may have introduced slower performance on some hardware. Or if one version is using line smoothing and one not.

I'll try and investigate, I just have a couple more crucial bugs to fix first (or clayjohn or anyone interested, it's not a massive amount of code to compare).

**Comment by lawnjelly** [comment]
If anyone getting this drop can try out 3.2.4 beta 2, and try these settings and see if there is any effect, that will be useful:
https://github.com/godotengine/godot/issues/43038#issuecomment-722336615

**Comment by pycbouh** [comment]
> If anyone getting this drop can try out 3.2.4 beta 2, and try these settings and see if there is any effect, that will be useful:
> [#43038 (comment)](https://github.com/godotengine/godot/issues/43038#issuecomment-722336615)

Tested. Only turning `rendering->options->api_usage_legacy->flag_stream` on has any effect, but it's a positive one. I get an fps boost even compared to 3.2.3, over 410 fps on average in the same sample project. Frame times are also better, giving constant stream of 2.3ms

![image](https://user-images.githubusercontent.com/11782833/114457785-19f26a00-9be7-11eb-82df-97b1d90bc245.png)

**Comment by lawnjelly** [comment]
See https://github.com/godotengine/godot/issues/47833#issuecomment-818506326 in the other issue, as although the observed results are slightly different, the underlying problems / discussion are similar.

**Comment by akien-mga** [comment]
I can confirm @pycbouh's results with the new options in #47864, turning ON `rendering/2d/legacy_stream` makes performance go from 80 FPS to 315 FPS (compared with 3.2.3-stable which had 188 FPS).

**Comment by lawnjelly** [comment]
> I can confirm @pycbouh's results with the new options in #47864, turning ON `rendering/2d/legacy_stream` makes performance go from 80 FPS to 315 FPS (compared with 3.2.3-stable which had 188 FPS).

Out of interest what are the results of this one with legacy_orphan set to off, and legacy_stream set to off?

**Comment by pycbouh** [comment]
I think I've tried all combinations between each pair of settings from 3.2.4b2 and nothing else provided any meaningful difference. Including both set to off.

**Comment by akien-mga** [comment]
> I can confirm @pycbouh's results with the new options in #47864, turning ON `rendering/2d/legacy_stream` makes performance go from 80 FPS to 315 FPS (compared with 3.2.3-stable which had 188 FPS).

Testing with the same device (Radeon RX Vega 56) on Linux, I get:

- Default:

```
Project FPS: 876 (1.1 mspf)
Project FPS: 874 (1.1 mspf)
Project FPS: 857 (1.1 mspf)
Project FPS: 831 (1.2 mspf)
Project FPS: 801 (1.2 mspf)
Project FPS: 869 (1.1 mspf)
Project FPS: 926 (1.0 mspf)
Project FPS: 831 (1.2 mspf)
Project FPS: 787 (1.2 mspf)
Project FPS: 789 (1.2 mspf)
Project FPS: 840 (1.1 mspf)
Project FPS: 842 (1.1 mspf)
```

- Legacy Stream ON:

```
Project FPS: 954 (1.0 mspf)
Project FPS: 922 (1.0 mspf)
Project FPS: 924 (1.0 mspf)
Project FPS: 946 (1.0 mspf)
Project FPS: 924 (1.0 mspf)
Project FPS: 928 (1.0 mspf)
Project FPS: 908 (1.1 mspf)
Project FPS: 930 (1.0 mspf)
Project FPS: 910 (1.0 mspf)
Project FPS: 847 (1.1 mspf)
Project FPS: 923 (1.0 mspf)
Project FPS: 911 (1.0 mspf)
```

Fluctuates a lot but I guess that can be summarized as a +100 FPS gain.

**Comment by lawnjelly** [comment]
I was just wondering because I hadn't yet worked out what had caused the drop in FPS from 3.2.3 to 3.3, as aside from the orphaning the code was mostly the same if I remember correctly. Maybe there is some other change since 3.2.3.

**Comment by pycbouh** [comment]
Something at least improves frame times in 3.3 overall. Not the average, maybe, but the spread. I definitely see it going off both ends in 3.2.4, but not so in 3.3 even with default/auto settings.

If there is anything else we can track, do tell.

**Comment by bruvzg** [comment]
Some tests all done on the same machine with AMD FirePro D300 GPU:

| OS / Driver   | Godot 3.2 | Godot 3.3-rc8 | Godot 3.x + 47864 and legacy options enabled |
| -- | -- | -- | -- |
| macOS 11.2.3 (20D91) | 105 FPS | 470 FPS | 460 FPS |
| Windows 10 (Radeon Software 21.3.1) | 230 FPS | 120 FPS | 210 FPS |
| Linux (Mesa 20.2.6/AMDGPU) | 270 FPS | 350 FPS | 350 FPS |

**Comment by lawnjelly** [comment]
As an aside, something I had noted a while ago is that the legacy `draw_polygon` and `draw_generic` etc functions are very inefficient API usage wise. (I think these are used to draw the circles).

They call `glBufferSubData` multiple times per call which could well be not good. I did make a bit of an attempt to improve this to a better method a while ago but made a mistake and reverted, but I've just not had it as a priority to fix. If we pre-prepared the whole buffer before uploading as a one off these calls would probably be a lot more efficient.

It's just not been a massive priority as I'm not sure how much these legacy primitives are used (they may be important in some editor modes though).

**Comment by pycbouh** [comment]
Out of curiousity I've also tried this project on my GTX 960m laptop. In 3.2.3 I get anywhere between 200 and 300 fps, sometimes dipping below 200. In 3.2.4 beta 2/3.3 RC 8 I get values between 35 and 50 fps, varying run to run. **Changing flags in 3.2.4 beta 2 doesn't have any effect**.

That's on Windows 10 and docked (not on battery). I didn't update either drivers or Windows on that laptop for some time though.

**Comment by bruvzg** [comment]
RC9 tests on the same machine (average over 1 minute of measurement, with first and last 5 seconds excluded), using official binaries for all tests, on Linux FPS have huge fluctuation for some reason:

| Godot version and config | macOS 11.2.3 (20D91) | Windows 10 (Radeon Software 21.3.1) | Ubuntu 20.10 (Mesa 20.2.6) |
| --- | --- | --- | --- |
| 3.3 RC9, `orphan_buffers=2`, `stream=1` | 450 ± 20 | 120 ± 10 | 465 ± 200 |
| 3.3 RC9, `orphan_buffers=1`, `stream=2` | 470 ± 20 | 205 ± 10 | 430 ± 200 |
| 3.3 RC9, `orphan_buffers=2`, `stream=2` | 445 ± 20 | 205 ± 10 | 460 ± 200 |
| 3.3 RC9, `orphan_buffers=1`, `stream=1` | 440 ± 20 | 120 ± 2 | 430 ± 100 |
| 3.2.3 stable | 110 ± 2 | 210 ± 2 | 265 ± 2 |

**Comment by lawnjelly** [comment]
Those results look encouraging, it should be good with the defaults we set in RC9.

The fluctuation on linux I'm not sure we can do anything obvious about, probably something you'd need internal knowledge of the driver to diagnose. Interesting that on windows the performance is halved with the same workload.

**Comment by akien-mga** [comment]
Fixed by #47864. Please comment if you still reproduce this specific issue in 3.3 RC 9 or stable.
